docs/HowToFAQ/FAQMapExtensionality.md

```diff
@@ -0,0 +1,114 @@
+Going back to your example, Dafny needs to know that the two `map`s are equal. To help it along, perhaps you could mention the two in an assertion, like
+
+`assert map … == map …;`
+
+But you can’t do that here, because the two map expressions are in different scopes and use different variables.
+To establish that the two maps are equal, we thus need to do something else. 
+If the two of them looked the same, then Dafny would know that the are the same. 
+But the form is slightly different, because you are (probably without thinking about it) instantiating a lambda expression. 
+To make them the same, you could rewrite the code to:
+```dafny
+  var F := k => Paths.SimpleCanon(tableName, k);
+  MapKeepsCount(authSchema, F);
+  assert |authSchema| == |map k <- authSchema.Keys | true :: F(k) := authSchema[k]|;
+```
+
+Now, this `map …` syntactically looks just like the one in the lemma postcondition, but with authSchema instead of m and with F instead of f. 
+When two map comprehensions (or set comprehensions, or lambda expressions) are syntactically the same like this, then Dafny knows to treat them the same.
+Almost. 
+Alas, there’s something about this example that makes what I said not quite true (and I didn’t dive into those details just now). 
+There is a workaround, and this workaround is often useful in places like this, so I’ll mention it here. 
+The workaround is to give the comprehension a name. Then, if you use the same name in both the caller and callee, 
+Dafny will know that they are the same way of constructing the map. 
+The following code shows how to do it: 
+```dafny
+lemma MapKeepsCount<X,Y,Z>(m : map<X,Y>, f : (X) -> Z)
+  requires forall a : X, b : X :: a != b ==> f(a) != f(b)
+  ensures |m| == |MyMap(f, m)|
+
+function MyMap<X, Y, Z>(f: X -> Y, m: map<X, Z>): map<Y, Z>
+  // (you can use the <- syntax in the following lines, by the Dafny Playground uses
+  // an older version of Dafny that didn't yet support that syntax)
+  requires forall a, b | a in m.Keys && b in m.Keys :: a != b ==> f(a) != f(b)
+{
+  // same comment about <- in the next line
+  map k | k in m.Keys :: f(k) := m[k]
+}
+
+method Use<X,Y,Z>(authSchema: map<X,Y>, tableName: Paths.TableName)
+  requires forall a : X, b : X :: a != b ==> Paths.SimpleCanon(tableName, a) != Paths.SimpleCanon(tableName, b)
+{
+  var F := k => Paths.SimpleCanon(tableName, k);
+  MapKeepsCount(authSchema, F);
+  assert |authSchema| == |MyMap(F, authSchema)|;
+}
+
+module Paths {
+  type TableName
+
+  function method SimpleCanon<K>(t: TableName, k: K): int
+}
+```
+
+It manually introduces a function MyMap, and by using it in both caller and callee, the code verifies.
+
+Being “injective” is the mathematical description of any function f that satisfies the property you’re giving in the precondition. 
+Another name for “injective” is “one-to-one” (which sounds more symmetric than it is--hence, the name “injective” is probably better).
```

Comment:
I don't think there's any reason to mention "injectivity" and "one-to-one" in this answer, since the question and the rest of the answer don't make use of these definitions.

Reply:
Removed